.gitlab-ci.yml

```diff
@@ -192,6 +192,35 @@ build:aws-k8s-pipeline-toolbox:
         --push
         aws-k8s-toolbox
 
+build:mender-dist-packages-image:
+  stage: build
+  needs: []
+  image: docker:git
+  services:
+    - docker:dind
+  script:
+    - CONTAINER_TAG=mender-dist-packages-builder-${DISTRO}-${RELEASE}-${ARCH}
+    - echo "INFO - Building and Pushing ${CONTAINER_TAG}-${CI_PIPELINE_ID} to the registry ${CI_REGISTRY_IMAGE}"
+    - echo $CI_REGISTRY_PASSWORD | docker login -u $CI_REGISTRY_USER $CI_REGISTRY --password-stdin
+    - docker build
+        --cache-from ${CI_REGISTRY_IMAGE}:${CONTAINER_TAG}-master
+        --tag ${CI_REGISTRY_IMAGE}:${CONTAINER_TAG}-${CI_PIPELINE_ID}
+        --build-arg DISTRO=${DISTRO}
+        --build-arg VERSION=${RELEASE}
+        --build-arg ARCH=${ARCH}
+        --build-arg GPG_KEY_BUILD="${GPG_KEY_BUILD}"
+        -f mender-dist-packages-building/Dockerfile
+        --push
+        mender-dist-packages-building
+  parallel:
+    matrix:
+      - DISTRO: debian
+        RELEASE: [buster, bullseye]
+        ARCH: [amd64, armhf, arm64]
+      - DISTRO: ubuntu
+        RELEASE: [bionic, focal, jammy]
+        ARCH: [amd64, armhf, arm64]
```

Comment:
just as a suggestion (since it popped up as my first search result [here](https://dnsmichi.at/2021/09/23/efficient-pipelines-gitlab-ci-cd-parallel-matrix-builds-reference/)) - moving this part to a template and referencing it in this job and the publish job should prevent the matrices from diverging...

Reply:
Thank you for the link, I didn't know about [reference tags](https://docs.gitlab.com/ee/ci/yaml/yaml_optimization.html#reference-tags) which has a lot of potential! I'll rework it.